File: lib/gameobjects/gameplayer.py
from ..database.db_regions import get_region
from ..database.db_startup_locations import get_startup_location

class GamePlayer(object):

    entered_game = False
    db_character = dict()
    race = ""
    ability_bonus = list()
    item_bonus = list()
    total_constitution_lost_at_death = -1
    # CharacterClass.ManaStat
    # CharacterClass.Name
    # CharacterClass.Profession
    level = 1
    max_health = -1
    concentration_effects = list()
    max_encumberance = -1
    encumberance = -1
    object_id = 1
    current_position = dict()
    current_zone = dict()
    current_region = dict()
    is_underwater = False
    guild = dict()
    has_horse, active_horse = False, dict()
    crafting_skills = list()
    is_turning_disabled = False
    max_speed = -1
    current_speed = -1
    money = dict()
    is_mezzed = False
    is_stunned = False
    is_strafing = False
    last_position_update_tick = -1
    last_position_update_point = {
        'X': -1,
        'Y': -1,
        'Z': -1
    }

    def __init__(self, db_character):
        self.db_character = db_character
        self.help_flag = self.db_character.get('no_help', True)
        self.init_current_region()
        self.init_current_position()
        self.init_money()

    def init_current_region(self):
        db_region = self.db_character['region']
        self.current_region = get_region(db_region)
        print self.current_region

    def init_current_position(self):
        self.current_position = {
            'X': self.db_character['x_pos'],
            'Y': self.db_character['y_pos'],
            'Z': self.db_character['z_pos'],
            'heading': 0
        }

        if self.current_position['X'] == 0 and self.current_position['Y'] == 0 and self.current_position['Z'] == 0:
            startup_location = get_startup_location(self.current_region['region_id'], self.db_character['realm'])
            if startup_location:
                self.current_position['X'] = startup_location['x_pos']
                self.current_position['Y'] = startup_location['y_pos']
                self.current_position['Z'] = startup_location['z_pos']
                self.current_position['heading'] = startup_location['heading']
        print self.current_position

    def init_money(self):
        self.money = {
            'copper': self.db_character['copper'],
            'silver': self.db_character['silver'],
            'gold': self.db_character['gold'],
            'mithril': self.db_character['mithril'],
            'platinium': self.db_character['platinium']
        }
        print self.money

File: src/main/scala/gameobjects/GamePlayer.scala
package gameobjects

import database._

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future
import scala.util.Random

/**
  * Created by franblas on 09/04/17.
  */
case class Position(x: Int, y: Int, z: Int, heading: Int)
case class Money(copper: Int, silver: Int, gold: Int, platinium: Int, mithril: Int)

class GamePlayer(charData: Character) {

  val region = new Regions()
  val startupLocation = new StartupLocations()
  val zone = new Zones()

  var enteredGame = false
  var dbCharacter: Character = _
  var race = ""
  //ability_bonus = list()
  //item_bonus = list()
  //total_constitution_lost_at_death = -1
  // CharacterClass.ManaStat
  // CharacterClass.Name
  // CharacterClass.Profession
  // level = 1
  // max_health = -1
  // concentration_effects = list()
  // max_encumberance = -1
  // encumberance = -1
  val objectId: Int = Random.nextInt(Short.MaxValue+1) //TODO
  var currentPosition: Position = _
  var currentZone: Zone = _
  var currentRegion: Region = _
  // is_underwater = False
  // guild = dict()
  // has_horse, active_horse = False, dict()
  // crafting_skills = list()
  // is_turning_disabled = False
  // max_speed = -1
  var currentSpeed: Int = 0
  var money: Money = _
  var isMezzed: Boolean = false
  var isStunned: Boolean = false
  var isStrafing: Boolean = false
  // last_position_update_tick = -1
  /*last_position_update_point = {
    'X': -1,
    'Y': -1,
    'Z': -1
  }*/

  def init(): Future[Unit] = {
    this.dbCharacter = charData
    initCurrentRegion().flatMap(result => {
      this.currentRegion = result.head
      initCurrentPosition()
    }).flatMap(_ => {
      initCurrentZone()
    }).map(_ => {
      initMoney()
    })
  }

  def initCurrentRegion(): Future[Seq[Region]] = {
    val regionId = dbCharacter.region
    region.getRegion(regionId)
  }

  def initCurrentPosition(): Future[Unit] = {
    this.currentPosition = Position(this.dbCharacter.x, this.dbCharacter.y, this.dbCharacter.z, 0)

    Future {
      if (this.currentPosition.x == 0 || this.currentPosition.y == 0 || this.currentPosition.z == 0) {
        startupLocation.getStartupLocation(this.currentRegion.regionId, this.dbCharacter.realm).map(result => {
          if (result.nonEmpty) {
            val startupLocation = result.head
            this.currentPosition = Position(
              startupLocation.x,
              startupLocation.y,
              startupLocation.z,
              startupLocation.heading
            )
          }
        })
      }
    }
  }

  def updateCurrentPosition(x: Int, y: Int, z: Int): Unit = {
    this.currentPosition = Position(x, y, z, this.currentPosition.heading)
  }

  def inZone(x: Int, y: Int, zone: Zone): Boolean = {
    val offsetX = zone.offsetX
    val offsetY = zone.offsetY
    val width = zone.width
    val height = zone.height
    val startX = 8192*offsetX
    val startY = 8192*offsetY
    val endX = startX + height * 8192
    val endY = startY + width * 8192
    if ((startX <= x && x <= endX) && (startY <= y && y <= endY)) true
    else false
  }

  def initCurrentZone(): Future[Unit] = {
    zone.getZoneFromRegion(this.currentRegion.regionId).map(result => {
      result.foreach(zone => {
        if (inZone(this.currentPosition.x, this.currentPosition.y, zone)) {
          this.currentZone = zone
        }
      })
    })
  }

  def initMoney() = {
    this.money = Money(
      this.dbCharacter.copper,
      this.dbCharacter.silver,
      this.dbCharacter.gold,
      this.dbCharacter.platinium,
      this.dbCharacter.mithril
    )
  }

}
